1. Simulate any random rectangular matrix A.
_usar numpy.rand_
- What is the rank and trace of A? 
_realizar explicacion sencilla_
- What is the determinant of A? 
- Can you invert A? How?
_para invertir el det debe ser 0_
_investigar sobre la pseudoinversa_
- How are eigenvalues and eigenvectors of A’A and AA’ related? What interesting differences can you notice between both?
_mostrar en código_ _eigenvalores son los mismos, eigenvectores son distintos_
- See [https://sites.cs.ucsb.edu/~mturk/Papers/jcn.pdf]


In [13]:
import pandas as pd
import numpy as np

In [14]:
A = np.random.rand(4,2)
A

array([[0.11327788, 0.60687846],
       [0.91288359, 0.02050799],
       [0.3339963 , 0.01876924],
       [0.47735734, 0.87307795]])

The rank of a matrix is the maximum number of columns or rows that are linearly independent. It's a measure of the linear independence of the columns or rows of a matrix. The rank of matrix `A` is 2 because is the smaller number between the dimensions of the matrix, which is 4 and 2. 

The trace is the sum along the diagonal of the matrix. The matrix `A` is not squared, but its possible to calculate the trace by adding the elements where row index and column index are equal. For matrix `A` the only elements that satisfy this condition are A(1,1) and A(2,2), so performing the addition of this elements results in the trace:

$Trace(A) = 0.11327788 + 0.02050799 = 0.1337858686554565$

In [33]:
# rank
rank = np.linalg.matrix_rank(A)
print(f"Rank of matrix A is {rank}.")
trace = np.trace(A)
print(f"Trace of matrix A is {trace}.")

Rank of matrix A is 2.
Trace of matrix A is 0.1337858686554565.


In [16]:
#transpose
print(f"shape of matrix is {A.shape}")
print(f"shape of transpose matrix is {A.T.shape}")

shape of matrix is (4, 2)
shape of transpose matrix is (2, 4)


In [15]:
# determinant
det = np.linalg.det(A)
det

LinAlgError: Last 2 dimensions of the array must be square

It's not possible to calculate determinant of `matrix` because its not an squared matrix. When `linalg.det` is called, an exception is thrown: _Last 2 dimensions of the array must be square_. Because of that, its also not possible to calculate the inverse of `matrix` having in mind that for a matrix to have inverse is required a determinant of 0. 

However, its possible to calculate the pseudo-inverse of non-square matrix. The pseudo-inverse matrix is like an 'approximation' of the inverse and its useful to solve linear equations, among other things. The notation for the pseudo-inverse matrix of matrix A is A^+ . To calculate the pseudo-inverse, proceed as follows:

$ A^{+} = (A^T A)^{-1}A^T $

In [ ]:
# inverse
inv = np.linalg.pinv(A)
inv

array([[ 0.95711356, -0.51576318,  0.20670102,  0.5798314 ],
       [-0.30586944,  0.67811312,  0.39155696, -0.04062271]])

In [30]:
# eig1 is A*A'
eigval1, eigvec1 = np.linalg.eig(np.dot(A, A.T))
# eig2 is A'*A
eigval2, eigvec2 = np.linalg.eig(np.dot(A.T, A))


print(f"eigenvalue of A*A' \n{eigval1}")
print(f"eigenvalue of A'*A \n{eigval2}")

print(f"eigenvalue of A*A' \n{eigvec1}")
print(f"eigenvalue of A'*A \n{eigvec2}")

eigenvalue of A*A' 
[1.66970271e+00+0.00000000e+00j 6.47248616e-01+0.00000000e+00j
 2.04127850e-17+1.66398108e-17j 2.04127850e-17-1.66398108e-17j]
eigenvalue of A'*A 
[1.66970271 0.64724862]
eigenvalue of A*A' 
[[-3.86776849e-01+0.j          4.50487442e-01+0.j
  -1.60190143e-02-0.31036662j -1.60190143e-02+0.31036662j]
 [-5.23559544e-01+0.j         -7.62269742e-01+0.j
  -3.15064010e-01-0.07522229j -3.15064010e-01+0.07522229j]
 [-1.97553664e-01+0.j         -2.68729965e-01+0.j
   8.66708986e-01+0.j          8.66708986e-01-0.j        ]
 [-7.32981325e-01+0.j          3.79196665e-01+0.j
  -9.68402178e-05+0.21750346j -9.68402178e-05-0.21750346j]]
eigenvalue of A'*A 
[[ 0.72563096 -0.68808409]
 [ 0.68808409  0.72563096]]


It can be seen that there are 2 eigenvalues for $A*A'$ and 4 eigenvalues for $A'*A$. Though the ammount of eigenvalues differs, the first 2 eigenvalues (1.66970271 and 0.64724862) are the same for both matrices $A*A'$ and $A'*A$. On the other hand, the eigenvectors are completely different for both matrices.

2. Add a steady, well-centered picture of your face to a shared folder alongside your classmates.
Edit your picture to be 256x256 pixels, grayscale (single channel)
Plot your edited face
_subir en teams una foto de la cara, centrada_
_usar matplotlib_
Calculate and plot the average face of the cohort
_usar numpy_
How distant is your face from the average? How would you measure it?
_como se calcula la distancia entre matrices_
_usar numpy frobenius distancia_
_


3. Let’s create the unsupervised Python package
_crear paquete, con clases, fit, fit_transform, hyperparms_
- Same API as scikit-learn: fit(), fit_transform(), transform(), hyperparams at init
- Manage dependencies with Pipenv or Poetry
- Implement SVD from scratch using Python and NumPy
- Implement PCA from scratch using Python and NumPy
[https://github.com/rushter/MLAlgorithms/blob/master/mla/pca.py,
https://github.com/patchy631/machine-learning/blob/main/ml_from_scratch/PCA_from_scratch.ip
ynb]
- Implement t-SNE from scratch using Python and NumPy
[https://nlml.github.io/in-raw-numpy/in-raw-numpy-t-sne/]


4. Apply SVD over the picture of your face, progressively increasing the number of singular values used. Is there any point where you can say the image is appropriately reproduced? How would you quantify how different your photo and the approximation are?
_jugar con la reconstrucción, cambio de componentes_


5. Train a naive logistic regression on raw MNIST images to distinguish between 0s and 8s. We are calling this our baseline. What can you tell about the baseline performance?
_usando scikit learn_

In [14]:
import pandas as pd

In [13]:
import sklearn

from sklearn.datasets import load_digits
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score 


In [21]:
mnist = load_digits(return_X_y=False)

In [22]:
mnist

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [24]:
Y = pd.DataFrame(mnist.target)
Y_sub = Y[Y[0].isin([0,8])]
Y_sub

0
0     0
8     8
10    0
18    8
20    0
...  ..
1789  8
1790  8
1793  0
1794  8
1796  8

[352 rows x 1 columns]

In [28]:
X = pd.DataFrame(mnist.data)
X

0    1     2     3     4     5    6    7    8    9   ...   54   55  \
0     0.0  0.0   5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1     0.0  0.0   0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
2     0.0  0.0   0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0   
3     0.0  0.0   7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0   
4     0.0  0.0   0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
...   ...  ...   ...   ...   ...   ...  ...  ...  ...  ...  ...  ...  ...   
1792  0.0  0.0   4.0  10.0  13.0   6.0  0.0  0.0  0.0  1.0  ...  4.0  0.0   
1793  0.0  0.0   6.0  16.0  13.0  11.0  1.0  0.0  0.0  0.0  ...  1.0  0.0   
1794  0.0  0.0   1.0  11.0  15.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1795  0.0  0.0   2.0  10.0   7.0   0.0  0.0  0.0  0.0  0.0  ...  2.0  0.0   
1796  0.0  0.0  10.0  14.0   8.0   1.0  0.0  0.0  0.0  2.0  ...  8.0  0.0   

       56   57   58    59    60    61   62   63  
0     0.0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1     0.0  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2     0.0  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3     0.0  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4     0.0  0.0  0.0   2.0  16.0   4.0  0.0  0.0  
...   ...  ...  ...   ...   ...   ...  ...  ...  
1792  0.0  0.0  2.0  14.0  15.0   9.0  0.0  0.0  
1793  0.0  0.0  6.0  16.0  14.0   6.0  0.0  0.0  
1794  0.0  0.0  2.0   9.0  13.0   6.0  0.0  0.0  
1795  0.0  0.0  5.0  12.0  16.0  12.0  0.0  0.0  
1796  0.0  1.0  8.0  12.0  14.0  12.0  1.0  0.0  

[1797 rows x 64 columns]

In [32]:
X_sub = X.iloc[Y_sub.index]
X_sub

0    1     2     3     4     5    6    7    8    9   ...   54   55  \
0     0.0  0.0   5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8     0.0  0.0   9.0  14.0   8.0   1.0  0.0  0.0  0.0  0.0  ...  8.0  0.0   
10    0.0  0.0   1.0   9.0  15.0  11.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
18    0.0  0.0  10.0   7.0  13.0   9.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
20    0.0  0.0   3.0  13.0  11.0   7.0  0.0  0.0  0.0  0.0  ...  1.0  0.0   
...   ...  ...   ...   ...   ...   ...  ...  ...  ...  ...  ...  ...  ...   
1789  0.0  0.0   8.0  16.0   3.0   0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0   
1790  0.0  0.0   5.0  12.0   8.0   0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0   
1793  0.0  0.0   6.0  16.0  13.0  11.0  1.0  0.0  0.0  0.0  ...  1.0  0.0   
1794  0.0  0.0   1.0  11.0  15.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1796  0.0  0.0  10.0  14.0   8.0   1.0  0.0  0.0  0.0  2.0  ...  8.0  0.0   

       56   57    58    59    60    61   62   63  
0     0.0  0.0   6.0  13.0  10.0   0.0  0.0  0.0  
8     0.0  0.0  11.0  16.0  15.0  11.0  1.0  0.0  
10    0.0  0.0   1.0  10.0  13.0   3.0  0.0  0.0  
18    0.0  0.0  11.0  14.0   5.0   0.0  0.0  0.0  
20    0.0  0.0   2.0  12.0  13.0   4.0  0.0  0.0  
...   ...  ...   ...   ...   ...   ...  ...  ...  
1789  0.0  0.0  10.0  16.0  10.0   1.0  0.0  0.0  
1790  0.0  0.0   6.0  12.0  12.0   3.0  0.0  0.0  
1793  0.0  0.0   6.0  16.0  14.0   6.0  0.0  0.0  
1794  0.0  0.0   2.0   9.0  13.0   6.0  0.0  0.0  
1796  0.0  1.0   8.0  12.0  14.0  12.0  1.0  0.0  

[352 rows x 64 columns]

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X_sub, Y_sub, test_size=0.33, random_state=42)

In [37]:
y_train = y_train.squeeze()

In [34]:
model = LogisticRegression()
model

LogisticRegression()

In [38]:
model.fit(X_train, y_train)

LogisticRegression()

In [40]:
y_pred = model.predict(X_test)
y_pred

array([0, 0, 8, 0, 0, 0, 0, 0, 8, 0, 8, 0, 8, 8, 8, 0, 8, 8, 0, 0, 0, 8,
       8, 0, 8, 8, 0, 8, 0, 0, 8, 8, 8, 8, 0, 0, 0, 0, 0, 0, 8, 0, 8, 8,
       8, 0, 8, 0, 0, 8, 8, 8, 8, 0, 8, 0, 0, 8, 0, 8, 8, 0, 0, 0, 8, 0,
       0, 0, 8, 0, 0, 8, 0, 8, 0, 0, 8, 0, 0, 8, 8, 8, 0, 8, 0, 8, 8, 8,
       8, 0, 8, 0, 0, 0, 8, 8, 0, 8, 8, 8, 8, 8, 0, 8, 8, 0, 0, 0, 8, 8,
       8, 8, 8, 8, 8, 0, 0])

In [41]:
accuracy_score(y_test, y_pred)

1.0

6. Now, apply dimensionality reduction using all your algorithms to train the model with only 2 features per image.
_ncomponents en 2_
- Plot the 2 new features generated by your algorithm
_usar matplotlib_
- Does this somehow impact the performance of your model?

7. Repeat the process above but now using the built-in algorithms in the Scikit-Learn library. How different are these results from those of your implementation? Why?

8. What strategies do you know (or can think of) in order to make PCA more robust? (Bonus points for implementing them)
[https://nbviewer.org/github/fastai/numerical-linear-algebra/blob/master/nbs/3.%20Background%20Remov
al%20with%20Robust%20PCA.ipynb]
_mas robusto a la escala, a la estandarización, buscar en google_

9. What are the underlying mathematical principles behind UMAP? What is it useful for?

10. What are the underlying mathematical principles behind LDA? What is it useful for?

11. Use your unsupervised Python package as a basis to build an HTTP server that receives a record as input and returns the class of the image. Suggestions: MNIST digit classifier, Iris classifier…
_ flask o fast api_
